File: engine/__tests__/pkSimulation.test.ts
import { getDrug } from '../switchingEngine';
import {
  effectiveHalfLifeHours,
  simulateDailyLevels,
  simulateSwitch,
} from '../pkSimulation';

describe('effectiveHalfLifeHours', () => {
  test('uses parent half-life when no clinically significant metabolite', () => {
    const sertraline = getDrug('sertraline')!;
    // Sertraline's metabolite is not clinically significant, so parent only.
    expect(effectiveHalfLifeHours(sertraline)).toBe(
      sertraline.halfLife.meanHours,
    );
  });

  test('extends half-life when active metabolite is clinically significant (fluoxetine)', () => {
    const fluoxetine = getDrug('fluoxetine')!;
    // Fluoxetine has norfluoxetine, clinically significant. Effective
    // half-life should be substantially longer than parent's 72h.
    const eff = effectiveHalfLifeHours(fluoxetine);
    expect(eff).toBeGreaterThan(fluoxetine.halfLife.meanHours);
    // norfluoxetine ~240h × 0.7 = 168h, larger than parent's 72h.
    expect(eff).toBe(168);
  });
});

describe('simulateDailyLevels', () => {
  test('starts at day-1 prescribed dose (assumes steady state at start)', () => {
    const points = [
      { day: 1, doseMg: 100 },
      { day: 4, doseMg: 50 },
    ];
    const sim = simulateDailyLevels(points, 26, 14);
    expect(sim[0]).toEqual({
      day: 1,
      prescribedDoseMg: 100,
      effectiveLevelMg: 100,
    });
  });

  test('effective level decays exponentially toward 0 after drug stopped', () => {
    // If prescribed drops to 0, the effective should follow the half-life curve.
    const points = [
      { day: 1, doseMg: 100 },
      { day: 2, doseMg: 0 },
    ];
    const sim = simulateDailyLevels(points, 24, 10); // exactly 1-day half-life
    // After 1 day at 0 (day 2): effective ≈ 100 * 0.5 + 0 * 0.5 = 50
    expect(sim[1].effectiveLevelMg).toBeCloseTo(50, 1);
    // Day 3: 50 * 0.5 = 25
    expect(sim[2].effectiveLevelMg).toBeCloseTo(25, 1);
    // Day 5: ~6.25
    expect(sim[4].effectiveLevelMg).toBeLessThan(10);
  });

  test('long-half-life drug shows long washout tail (fluoxetine)', () => {
    const fluoxetine = getDrug('fluoxetine')!;
    const points = [
      { day: 1, doseMg: 20 },
      { day: 4, doseMg: 0 }, // stop fluoxetine
    ];
    const sim = simulateDailyLevels(
      points,
      effectiveHalfLifeHours(fluoxetine),
      30,
    );
    // 14 days after stopping (day 18), level should still be measurably above 0
    // because effective half-life is 168h = 7 days.
    const day18 = sim.find((p) => p.day === 18)!;
    expect(day18.effectiveLevelMg).toBeGreaterThan(2); // still ~25% of original 20mg
  });
});

describe('simulateSwitch', () => {
  test('returns matched-length from and to series with trailing days', () => {
    const fromDrug = getDrug('sertraline')!;
    const toDrug = getDrug('escitalopram')!;
    const schedule = [
      { day: 1, fromDoseMg: 100, toDoseMg: 5 },
      { day: 11, fromDoseMg: 0, toDoseMg: 10 },
    ];
    const sim = simulateSwitch(schedule, fromDrug, toDrug);
    // Default trailing 14 days => 11 + 14 = 25
    expect(sim.totalDays).toBe(25);
    expect(sim.from.length).toBe(25);
    expect(sim.to.length).toBe(25);
    // From-drug effective decays after day 11
    expect(sim.from[24].effectiveLevelMg).toBeLessThan(sim.from[0].effectiveLevelMg);
    // To-drug effective rises toward target
    expect(sim.to[24].effectiveLevelMg).toBeGreaterThan(sim.to[0].effectiveLevelMg);
  });
});

File: flutter_app/test/engine/pk_simulation_test.dart
// Tests for the Dart pk_simulation port.
// Mirrors engine/__tests__/pkSimulation.test.ts but uses inline Drug
// fixtures rather than loading from /content/drugs/. The values match
// the relevant fields in the canonical JSON files.

import 'package:flutter_test/flutter_test.dart';
import 'package:psychswitch/src/engine/pk_simulation.dart';
import 'package:psychswitch/src/engine/types/drug.dart';
import 'package:psychswitch/src/engine/types/schedule_step.dart';

Drug _drug({
  required String id,
  required String name,
  required double parentHalfLifeH,
  ActiveMetabolite? metabolite,
}) =>
    Drug(
      id: id,
      genericName: name,
      drugClass: 'antidepressant',
      malaysianBrandNames: const <String>[],
      halfLife: HalfLife(
        meanHours: parentHalfLifeH,
        rangeHours: const <double>[],
      ),
      activeMetabolite: metabolite ??
          const ActiveMetabolite(
            name: null,
            halfLifeHours: null,
            clinicallySignificant: false,
          ),
      cypInteractions: const CypInteractions(
        substrateOf: <String>[],
        inhibitorOf: <String>[],
        switchingRelevance: '',
      ),
      dosing: const Dosing(
        startingDoseMg: 0,
        typicalTargetRangeMg: <double>[],
        maxDoseMg: 0,
        increments: <double>[],
        formulationsAvailableMy: <String>[],
      ),
      formulationNotes: '',
      citations: const <String>[],
      lastReviewedISO: '2026-04-01',
      reviewedBy: 'Test',
    );

void main() {
  group('effectiveHalfLifeHours', () {
    test('uses parent half-life when no clinically significant metabolite',
        () {
      final sertraline = _drug(
        id: 'sertraline',
        name: 'Sertraline',
        parentHalfLifeH: 26,
      );
      expect(effectiveHalfLifeHours(sertraline), equals(26));
    });

    test(
      'extends half-life when active metabolite is clinically significant '
      '(fluoxetine)',
      () {
        final fluoxetine = _drug(
          id: 'fluoxetine',
          name: 'Fluoxetine',
          parentHalfLifeH: 72,
          metabolite: const ActiveMetabolite(
            name: 'Norfluoxetine',
            halfLifeHours: 240,
            clinicallySignificant: true,
          ),
        );
        final eff = effectiveHalfLifeHours(fluoxetine);
        expect(eff, greaterThan(72));
        expect(eff, equals(168));
      },
    );
  });

  group('simulateDailyLevels', () {
    test('starts at day-1 prescribed dose (assumes steady state at start)',
        () {
      const points = <SchedulePoint>[
        SchedulePoint(day: 1, doseMg: 100),
        SchedulePoint(day: 4, doseMg: 50),
      ];
      final sim = simulateDailyLevels(points, 26, 14);
      expect(sim[0].day, equals(1));
      expect(sim[0].prescribedDoseMg, equals(100));
      expect(sim[0].effectiveLevelMg, equals(100));
    });

    test('effective level decays exponentially toward 0 after drug stopped',
        () {
      const points = <SchedulePoint>[
        SchedulePoint(day: 1, doseMg: 100),
        SchedulePoint(day: 2, doseMg: 0),
      ];
      final sim = simulateDailyLevels(points, 24, 10);
      expect(sim[1].effectiveLevelMg, closeTo(50, 0.1));
      expect(sim[2].effectiveLevelMg, closeTo(25, 0.1));
      expect(sim[4].effectiveLevelMg, lessThan(10));
    });

    test('long-half-life drug shows long washout tail (fluoxetine)', () {
      final fluoxetine = _drug(
        id: 'fluoxetine',
        name: 'Fluoxetine',
        parentHalfLifeH: 72,
        metabolite: const ActiveMetabolite(
          name: 'Norfluoxetine',
          halfLifeHours: 240,
          clinicallySignificant: true,
        ),
      );
      const points = <SchedulePoint>[
        SchedulePoint(day: 1, doseMg: 20),
        SchedulePoint(day: 4, doseMg: 0),
      ];
      final sim = simulateDailyLevels(
        points,
        effectiveHalfLifeHours(fluoxetine),
        30,
      );
      final day18 = sim.firstWhere((p) => p.day == 18);
      expect(day18.effectiveLevelMg, greaterThan(2));
    });

    test('returns empty for empty schedule', () {
      expect(simulateDailyLevels(<SchedulePoint>[], 24, 5), isEmpty);
    });
  });

  group('simulateSwitch', () {
    test('returns matched-length from and to series with trailing days', () {
      final fromDrug = _drug(
        id: 'sertraline',
        name: 'Sertraline',
        parentHalfLifeH: 26,
      );
      final toDrug = _drug(
        id: 'escitalopram',
        name: 'Escitalopram',
        parentHalfLifeH: 30,
      );
      final schedule = <ScheduleStep>[
        const ScheduleStep(day: 1, fromDoseMg: 100, toDoseMg: 5),
        const ScheduleStep(day: 11, fromDoseMg: 0, toDoseMg: 10),
      ];
      final sim = simulateSwitch(schedule, fromDrug, toDrug);
      expect(sim.totalDays, equals(25));
      expect(sim.from.length, equals(25));
      expect(sim.to.length, equals(25));
      expect(
        sim.from.last.effectiveLevelMg,
        lessThan(sim.from.first.effectiveLevelMg),
      );
      expect(
        sim.to.last.effectiveLevelMg,
        greaterThan(sim.to.first.effectiveLevelMg),
      );
    });

    test('honours custom trailingDays', () {
      final d = _drug(id: 'x', name: 'X', parentHalfLifeH: 24);
      final schedule = <ScheduleStep>[
        const ScheduleStep(day: 1, fromDoseMg: 50, toDoseMg: 0),
        const ScheduleStep(day: 5, fromDoseMg: 0, toDoseMg: 25),
      ];
      final sim = simulateSwitch(schedule, d, d, trailingDays: 3);
      expect(sim.totalDays, equals(8));
    });
  });
}
